### PR TITLE
[WIP][RFT] generic: platform/mikrotik: add wlan lz77 decompress

### PR DESCRIPTION
**Pull request by john-tho**
usual disclaimer: expect your device to halt and catch fire…

very rough attempt to decode lz77; rough, excessively verbose, and entirely unoptimised.

A number of new (or newly factory flashed) Mikrotik devices are using LZ77 magic for wlan tag hard_config data.
New devices include the Chateau LTE12 [1], and ax devices [2] Newly factory flashed devices may include the hap ac3 [3]

This can be seen in decoded OEM supout [4] dmesg:
"radio data lz77 decompressed from"…

Investigating an arm RouterOS flash.ko module, and supplied example hard_config dumps, the format was determined via [5].

[1]: https://forum.openwrt.org/t/no-wireless-mikrotik-rbd53ig-5hacd2hnd/157763/4
[2]: https://forum.openwrt.org/t/mikrotik-routeros-v7-x-and-openwrt-sysupgrade/148072/17
[3]: https://forum.openwrt.org/t/adding-support-for-mikrotik-hap-ax2/133715/47
[4]: https://github.com/farseeker/go-mikrotik-rif
[5]: https://github.com/john-tho/routeros-wlan-lz77-decode

### CODE REVIEW

File: target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig_lz77.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+		unsigned char *out,
+		size_t *out_len);
+#else
+static inline int lz77_mikrotik_wlan_decompress(
+		const unsigned char *in,
+		size_t in_len,
+		unsigned char *out,
+		size_t *out_len)
+{
+	return -ENOTSUP;
```
Comment (john-tho) [inline target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig_lz77.h:34]:
```suggestion
	return -ENOTSUPP;
```

File: target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig_lz77.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#define LZ77_MK_EXPECTED_OUT 0x44524500
+
+/* the look behind window
+ * unknown, for long instruction match offsets up to
+ * 6449 have been seen (would need 21 counter bits: 4 to 12 + 11 to 0)
+ * conservative value here: 27 provides offset up to 0x8000 bytes
+ */
+#define LZ77_MK_MAX_COUNT_BIT_LEN 27
+
+#ifdef CONFIG_CPU_LITTLE_ENDIAN
```
Comment (john-tho) [inline target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig_lz77.h:21]:
```suggestion
#if defined (CONFIG_CPU_LITTLE_ENDIAN) || defined(CONFIG_ARCH_IPQ40XX)
```
`ipq4019` does not set `CONFIG_CPU_LITTLE_ENDIAN`

Comment (robimarko) [reply target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig_lz77.h]: I think it does in the final kernel config

**Comment by dchard** [comment]
The latest push is tested on Mikrotik Chateau-12LTE (ipq4019). Both radios are working, driver does load the firmware with BDF/CAL properly, WLAN MACs are also correct. Test is conducted on Routerboot_v7 with the required additional patches. Dmesg is completely fine, no errors or weird messages.

` firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5`

If you want you can use my "tested by" tag:
Csaba Sipos <metro4_at_freemail.hu>

**Comment by john-tho** [comment]
Hopefully no functional changes with push, only fixing tidy messages, and fix some checkpatch and sparse warnings

**Comment by dchard** [comment]
> Hopefully no functional changes with push, only fixing tidy messages, and fix some checkpatch and sparse warnings

I can recheck it early next week.

**Comment by dchard** [comment]
@john-tho found a device out of the many that fails at LZ77 decompress:

```
[    0.960517] [rb_hardconfig][lz77] input overrun
[    0.960565] [rb_hardconfig] LZ77: LZ77 decompress fail
[    0.964427] [rb_hardconfig][lz77] input overrun
[    0.969139] [rb_hardconfig] LZ77: LZ77 decompress fail
[    0.974058] [rb_hardconfig][lz77] input overrun
[    0.978735] [rb_hardconfig] LZ77: LZ77 decompress fail
[    0.983165] MikroTik RouterBOARD hardware configuration sysfs driver v0.08
[    0.992248] MikroTik RouterBOARD software configuration sysfs driver v0.05
```
And this is the MTD2 (hard config) dump of the same device:

[LZ77_decomp_failed_mtd2_hard_config.zip](https://github.com/john-tho/openwrt/files/12099277/LZ77_decomp_failed_mtd2_hard_config.zip)

**Comment by john-tho** [comment]
> @john-tho found a device out of the many that fails at LZ77 decompress:
> 
> ```
> [    0.960517] [rb_hardconfig][lz77] input overrun
> [    0.960565] [rb_hardconfig] LZ77: LZ77 decompress fail
> [    0.964427] [rb_hardconfig][lz77] input overrun
> [    0.969139] [rb_hardconfig] LZ77: LZ77 decompress fail
> [    0.974058] [rb_hardconfig][lz77] input overrun
> [    0.978735] [rb_hardconfig] LZ77: LZ77 decompress fail
> [    0.983165] MikroTik RouterBOARD hardware configuration sysfs driver v0.08
> [    0.992248] MikroTik RouterBOARD software configuration sysfs driver v0.05
> ```
> 
> And this is the MTD2 (hard config) dump of the same device:
> 
> LZ77_decomp_failed_mtd2_hard_config.zip

Okay, thanks, I pulled a copy.

In my standalone decompressor, looks like the final group (non-match) bits have some set bits.
```
inbit:0x55d1->outbyte:0x2e8a bits opcode:0xc match, offset: 0x1, len: 0x6 (5 partial memcpy)
inbit:0x55dd->outbyte:0x2e90 bits opcode:0xc non-match, len: 0xc
 correct len
(input_bit 0x55e9 + opcode->length 0xc*8) = 0x5649 > in_len 0xac0 (bits 0x5600)
 correct end
input bit 0x55e9, input_len * 8: 0x5600
input byte -1 0x0, 0xe
input bit 0x55dd
opcode bits
 1 1 0 0 0 0 0 0 0 0 0 0
non-match data bits
 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
input overrun
finished, rc=-1
output first bytes:
44 52 45 00 01 82 28 19
```

You could try removing the `&& (!lz77_mikrotik_wlan_get_bit(in, input_bit))` from the `test end marker` section.
It might just work, and my null bit(s) follows last group opcode assumption (based on all prior examples) was wrong.

We should look to see how the Mikrotik binary handles this last group & payload.

**Comment by john-tho** [comment]
> In my standalone decompressor, looks like the final group (non-match) bits have some set bits.
> 
> ```
> inbit:0x55d1->outbyte:0x2e8a bits opcode:0xc match, offset: 0x1, len: 0x6 (5 partial memcpy)
> inbit:0x55dd->outbyte:0x2e90 bits opcode:0xc non-match, len: 0xc
>  correct len
> (input_bit 0x55e9 + opcode->length 0xc*8) = 0x5649 > in_len 0xac0 (bits 0x5600)
>  correct end
> input bit 0x55e9, input_len * 8: 0x5600
> input byte -1 0x0, 0xe
> input bit 0x55dd
> opcode bits
>  1 1 0 0 0 0 0 0 0 0 0 0
> non-match data bits
>  1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
> input overrun
> finished, rc=-1
> output first bytes:
> 44 52 45 00 01 82 28 19
> ```
> 
> You could try removing the `&& (!lz77_mikrotik_wlan_get_bit(in, input_bit))` from the `test end marker` section. It might just work, and my null bit(s) follows last group opcode assumption (based on all prior examples) was wrong.
> 
> We should look to see how the Mikrotik binary handles this last group & payload.

Emulated binary decompressed payload was `0x2e90` bytes long, so I should be able to cut the extra test for bits following the last group opcode: `&& (!lz77_mikrotik_wlan_get_bit(in, input_bit))`

**Comment by dchard** [comment]
@john-tho just to be sure: this does not looks like a potential flash/data corruption to you? I used your patch on about 80 devices from multiple batches of Chateau LTE12 and there was a single one so far which produced this fault.

**Comment by john-tho** [comment]
> @john-tho just to be sure: this does not looks like a potential flash/data corruption to you?

Does not look like corrupted flash. If it was, this lz77 decompress would fail very differently. Just extra set bits after the last (non) match group.

**Comment by dchard** [comment]
> > @john-tho just to be sure: this does not looks like a potential flash/data corruption to you?
> 
> Does not look like corrupted flash. If it was, this lz77 decompress would fail very differently. Just extra set bits after the last (non) match group.

Tested your latest modifications on the affected device, and now it works without any errors.

**Comment by aanon4** [comment]
How little endian is this decompression code? I have two ath79 based device (921GS-5HPacD-19s - one of their newer sectors) and one boots fine while the other doesn't load the firmware. I was thinking this compression scheme might be the problem there also? However this device is big endian so I wondered how portable it was?
Thanks

**Comment by john-tho** [comment]
> How little endian is this decompression code? I have two ath79 based device (921GS-5HPacD-19s - one of their newer sectors) and one boots fine while the other doesn't load the firmware. I was thinking this compression scheme might be the problem there also? However this device is big endian so I wondered how portable it was? Thanks
> 
> [Edit] Looking though the hard_config data, it is definitely LZ77 compressed.
> 
> [Edit 2] I tried it "as is" and it works great! Thanks so much. You might want to consider widening the scope of where this is included.

Ha! Colour me surprised. I was not expecting to see this used on MIPS. Many thanks for the new data and testing Tim. I guess, because the compressed data is read one bit at a time, that _should_ have been the only place where endian-ness mattered. If you could, I would appreciate you detailing the backup RouterBOOT version so we might be able to piece together some idea of when this started: `cat /sys/firmware/mikrotik/hard_config/booter_version`

Okay, I will adjust the `LE|IPQ40XX` build conditional next time I touch this.

Cheers.

**Comment by aanon4** [comment]
The output from that cat command:
```6.48.6```

**Comment by dchard** [comment]
@john-tho just checked the LZ77 and routeros-v7 patches on the new 6.1.55 kernel for IPQ4019 and it seems to work correctly.

**Comment by MichaelUray** [comment]
The patch works for me with a MikroTik hAP ac3.
What is actually required to get that merged into the master?

**Comment by omidomidomid** [comment]
any new update ? I got this problem with my hap ac3 and wireless interfaces wont detected. any suggestion ?

**Comment by robimarko** [comment]
@john-tho Any chance you are planning to make a PR for this?

**Comment by omidomidomid** [comment]
Thanks John
Is there any chance if you could provide the build result. The sysuograde verdion i mean. I am not sure if i can build the branch myself.

**Comment by john-tho** [comment]
Hi @aanon4,

> I have two ath79 based device (921GS-5HPacD-19s - one of their newer sectors) and one boots fine while the other doesn't load the firmware.
> [Edit] Looking though the hard_config data, it is definitely LZ77 compressed.

I am gradually getting this together for upstream OpenWrt (https://github.com/openwrt/openwrt/pull/15774), and have made a few small changes to the code. I would be interested to see if you could try out this updated code, and/or email me a copy of `hard_config` and `soft_config` mtd partitions for that lz77 ath79 device?

> The output from that cat command:
> `6.48.6`

There are a number of Mikrotik ath79 devices with ath10k pcie cards, and I have not noticed any other OpenWrt forum or issue reports of no 5GHz Wi-Fi on them (but for some time they have had more modern replacements). My guess now is that this may occur (for ath79) if calibration data is updated (guessing in RouterOS update) for US devices to enable U-NII 2 frequency ranges, or recent factory flashes.

**Comment by aanon4** [comment]
> Hi @aanon4,
> 
> > I have two ath79 based device (921GS-5HPacD-19s - one of their newer sectors) and one boots fine while the other doesn't load the firmware.
> > [Edit] Looking though the hard_config data, it is definitely LZ77 compressed.
> 
> I am gradually getting this together for upstream OpenWrt ([openwrt#15774](https://github.com/openwrt/openwrt/pull/15774)), and have made a few small changes to the code. I would be interested to see if you could try out this updated code, and/or email me a copy of `hard_config` and `soft_config` mtd partitions for that lz77 ath79 device?
> 
> > The output from that cat command:
> > `6.48.6`
> 
> There are a number of Mikrotik ath79 devices with ath10k pcie cards, and I have not noticed any other OpenWrt forum or issue reports of no 5GHz Wi-Fi on them (but for some time they have had more modern replacements). My guess now is that this may occur (for ath79) if calibration data is updated (guessing in RouterOS update) for US devices to enable U-NII 2 frequency ranges, or recent factory flashes.

@john-tho Unfortunately this is tricky. I work on the AREDN project which is built on top of OpenWRT. As such the radio in question is currently at the top of a tower at the top of a mountain. I can't really run test code on it because ... well ... getting to it if fails is a massive pain.

**Comment by john-tho** [comment]
> @john-tho Unfortunately this is tricky. I work on the AREDN project which is built on top of OpenWRT. As such the radio in question is currently at the top of a tower at the top of a mountain. I can't really run test code on it because ... well ... getting to it if fails is a massive pain.

Haha, yes, no worries, kind of expected this. I too have not-so-easily-accessible devices. Lab test devices are boring, field tests (or deploys) are much more interesting.
If someone could possible remote into the device, and pull a dump of the mtd hard_config partition, that would allow me to run some sanity checks on my code at least. soft_config could just tell us if RouterBoot primary had been updated as well.
Also, please let me know if you heard of any other ath79 devices that needed this (used LZ77 calibration data)?
Cheers.

**Comment by aanon4** [comment]
Noting this here despite this being merged already.
I realized yesterday that one of my Mikrotik 921GS-5HPacD-19s no longer detected its radio. This is running code based on OpenWRT 24.10 (latest release as I write this). Turned out there was a slight change between the patch set I originally got here and what's in the release. The diff, to make my node work again, is below.  I guess there is a "known instance" now.

--- a/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c
+++ b/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c
@@ -592,12 +592,6 @@
 	ret = -ENODATA;
 	switch (magic) {
 	case RB_MAGIC_LZ77:
-		/* no known instances of lz77 without 8001/8201 data, skip SOLO */
-		if (tag_id == RB_WLAN_ERD_ID_SOLO) {
-			pr_debug(RB_HC_PR_PFX "skipped LZ77 decompress in search for SOLO tag\n");
-			break;
-		}
-		fallthrough;
 	case RB_MAGIC_LZOR:
 		/* Skip magic */
 		lbuf += sizeof(magic);

**Comment by john-tho** [comment]
> Mikrotik 921GS-5HPacD-19s, to make my node work again
> 
> ```
> --- a/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c
> +++ b/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c
> @@ -592,12 +592,6 @@
>  	ret = -ENODATA;
>  	switch (magic) {
>  	case RB_MAGIC_LZ77:
> -		/* no known instances of lz77 without 8001/8201 data, skip SOLO */
> -		if (tag_id == RB_WLAN_ERD_ID_SOLO) {
> -			pr_debug(RB_HC_PR_PFX "skipped LZ77 decompress in search for SOLO tag\n");
> -			break;
> -		}
> -		fallthrough;
>  	case RB_MAGIC_LZOR:
>  		/* Skip magic */
>  		lbuf += sizeof(magic);
> ```

Hi Tim,
Ha, thank you. Good find. Care to submit this as a PR to OpenWrt master (then we cherry-pick backport to 24)?

**Pull request merged**